File: 0261-graph-valid-tree/0261-graph-valid-tree.js
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {boolean}
 */
var validTree = function(n, edges) {
    const adjList = new Array(n).fill().map(()=>[])
    for(const [start,end] of edges){
        adjList[start].push(end)
        adjList[end].push(start)
    }

    const visited = new Array(n).fill(0)
    const parent = new Array(n).fill(-1)

    function helper(node){
        visited[node] = 1
        for(const nei of adjList[node]){
            if(visited[nei] === 0){
                parent[nei] = node
                const sub = helper(nei) // check sub result
                if(sub === false){
                    return false
                }
            } else {
                if(parent[node] != nei){
                    return false
                }
            }
        }
        return true
    }

    let component = 0

    for(let i=0; i<n; i++){
        if(visited[i] === 0){
            const isTree = helper(i)
            component++
            if(component > 1 || !isTree){ // do early return
                return false
            }
        }
    }

    return true
    
};

File: 0261-graph-valid-tree/0261-graph-valid-tree.py
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        if len(edges) != n-1:
            return False

        adjList = defaultdict(list)
        for a, b in edges:
            adjList[a].append(b)
            adjList[b].append(a)
        
        visited = set()
        
        def dfs(node, parent):
            visited.add(node)
            for neigh in adjList[node]:
                if neigh == parent:
                    continue
                if neigh in visited:
                    return False
                if not dfs(neigh, node):
                    return False
            return True
        
        return dfs(0, -1) and len(visited) == n

